# Pregunta 5

In [10]:
import pandas as pd

# Importamos el conjunto de datos
boston_data = pd.read_csv('C:/Users/arica.MIJOCHA/Documents/GitHub/Qlab_Repositorio/Introducción a Machine Learning/PC2/boston.csv')


boston_data

Unnamed: 0     crim    zn  indus  chas    nox     rm   age     dis  rad  \
0             0  0.00632  18.0   2.31     0  0.538  6.575  65.2  4.0900    1   
1             1  0.02731   0.0   7.07     0  0.469  6.421  78.9  4.9671    2   
2             2  0.02729   0.0   7.07     0  0.469  7.185  61.1  4.9671    2   
3             3  0.03237   0.0   2.18     0  0.458  6.998  45.8  6.0622    3   
4             4  0.06905   0.0   2.18     0  0.458  7.147  54.2  6.0622    3   
..          ...      ...   ...    ...   ...    ...    ...   ...     ...  ...   
501         501  0.06263   0.0  11.93     0  0.573  6.593  69.1  2.4786    1   
502         502  0.04527   0.0  11.93     0  0.573  6.120  76.7  2.2875    1   
503         503  0.06076   0.0  11.93     0  0.573  6.976  91.0  2.1675    1   
504         504  0.10959   0.0  11.93     0  0.573  6.794  89.3  2.3889    1   
505         505  0.04741   0.0  11.93     0  0.573  6.030  80.8  2.5050    1   

     tax  ptratio  lstat  medv  
0    296     15.3   4.98  24.0  
1    242     17.8   9.14  21.6  
2    242     17.8   4.03  34.7  
3    222     18.7   2.94  33.4  
4    222     18.7   5.33  36.2  
..   ...      ...    ...   ...  
501  273     21.0   9.67  22.4  
502  273     21.0   9.08  20.6  
503  273     21.0   5.64  23.9  
504  273     21.0   6.48  22.0  
505  273     21.0   7.88  11.9  

[506 rows x 14 columns]

In [11]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import Ridge, Lasso
from sklearn.metrics import mean_squared_error
import numpy as np

# Preparamos los datos
X = boston_data.drop(['medv', 'Unnamed: 0'], axis=1)  # Eliminamos la columna objetivo y la columna no nombrada
y = boston_data['medv']

(23.007013892137614, 23.328630966752637)

Dividimos el conjunto de datos en un conjunto de entrenamiento y un conjunto de validación (también conocido como conjunto de prueba) usando la función train_test_split de Scikit-learn.

Esto asegura que la evaluación del rendimiento del modelo refleje su capacidad para trabajar con datos que no había visto antes, lo cual es una práctica estándar en el aprendizaje automático para evitar el sobreajuste y obtener una estimación realista del rendimiento del modelo.

In [13]:
# Dividimos los datos en conjuntos de entrenamiento y validación
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Función para entrenar y evaluar un modelo
def train_and_evaluate(model, X_train, X_test, y_train, y_test):
    model.fit(X_train, y_train)
    y_pred = model.predict(X_test)
    mse = mean_squared_error(y_test, y_pred)
    return mse

# Entrenamos y evaluamos el modelo Ridge
ridge_model = Ridge()
ridge_mse = train_and_evaluate(ridge_model, X_train, X_test, y_train, y_test)

# Entrenamos y evaluamos el modelo Lasso
lasso_model = Lasso()
lasso_mse = train_and_evaluate(lasso_model, X_train, X_test, y_train, y_test)

ridge_mse, lasso_mse

(23.007013892137614, 23.328630966752637)

# Resultados preliminares

✔️**Modelo Ridge:** El error cuadrático medio (MSE) en el conjunto de prueba es aproximadamente 23.01.

✔️**Modelo Lasso:** El MSE en el conjunto de prueba es aproximadamente 23.33.
Ambos modelos tienen un rendimiento similar en términos de MSE en el conjunto de prueba. Sin embargo, el modelo Ridge tiene un ligero mejor desempeño en este caso.

# Aplicación de modelo alternativo:

Comparé los modelos Ridge y Lasso con un modelo más sencillo, la regresión lineal. Sorprendentemente, el modelo más sencillo fue un poco mejor para predecir los precios de las casas. Esto sugiere que, para este conjunto de datos, no necesitamos métodos tan complejos como Ridge o Lasso.

In [12]:
from sklearn.linear_model import LinearRegression

# Entrenar y evaluar el modelo de regresión lineal
linear_model = LinearRegression()
linear_mse = train_and_evaluate(linear_model, X_train, X_test, y_train, y_test)

linear_mse

22.777708563865982

✔️El **modelo de regresión lineal** simple tiene un error cuadrático medio (MSE) de aproximadamente 22.78 en el conjunto de prueba, lo cual es ligeramente mejor que los modelos Ridge y Lasso.

# Conclusiones:

El modelo de regresión lineal que usamos tomó en cuenta todas las características del conjunto de datos, como el tamaño de las casas, la edad, etc. Esto fue porque, en nuestra prueba, usar todas estas características juntas nos dio los mejores resultados.

Aunque usamos todas las características, esto no significa que todas sean igualmente importantes. Sería bueno estudiar más a fondo cuáles son realmente importantes y cuáles no tanto. Esto podría hacer que nuestro modelo sea más sencillo y fácil de entender, pero en este caso, usar todas las características funcionó bien.




